packages/bug-detectors/internal/prototype-pollution.ts

```diff
@@ -0,0 +1,565 @@
+	// Compare the current prototype snapshots of basic objects to the original ones.
+	for (let i = 0; i < BASIC_PROTO_SNAPSHOTS.length; i++) {
+		if (!currentProtoSnapshots[i]) {
+			reportFinding(
+				`Prototype Pollution: prototype of ${BASIC_OBJECT_NAMES[i]} changed.`
+			);
+			continue;
+		}
+		const equalityResult = protoSnapshotsEqual(
+			BASIC_PROTO_SNAPSHOTS[i],
+			currentProtoSnapshots[i]
+		);
+		if (!equalityResult.equal) {
+			reportFinding(
+				`Prototype Pollution: prototype of ${BASIC_OBJECT_NAMES[i]} changed.${equalityResult.message}`
+			);
+		}
+	}
+}
+
+registerAfterEachCallback(detectPrototypePollutionOfBasicObjects);
+
+// There are two main ways to pollute a prototype of an object:
+// 1. Changing a prototype's property using __proto__
+// 2. Changing a prototype's property using constructor.prototype
+// This dictionary adds these strings to the fuzzer dictionary.
+// Adding strings targeted at specific protocols (XML, HTTP, protobuf, etc.) will reduce the performance of the fuzzer,
+// because it will try strings from the wrong protocol. Therefore, it is advised to add protocol-specific strings
+// to the user dictionary for each fuzz test individually.
+addDictionary("__proto__", "constructor", "prototype", "constructor.prototype");
+
+type BasicProtoSnapshots = ProtoSnapshot[];
+
+function computeBasicPrototypeSnapshots(): BasicProtoSnapshots {
+	return BASIC_OBJECTS.map(getProtoSnapshot);
+}
+
+/**
+ * Checks if the object's proto contains any non-function properties. Function properties are ignored.
+ * @param obj The object to check.
+ * @param identifier The identifier of the object (used for printing a useful finding message).
+ * @param report Whether to report a finding if the object is a prototype pollution object.
+ */
+function detectPrototypePollution(
+	// eslint-disable-next-line  @typescript-eslint/no-explicit-any
+	obj: any,
+	identifier?: string,
+	report = true
+): boolean | void {
+	while (obj || obj === false || obj === 0 || obj === "") {
+		let protoValue: string;
+		try {
+			protoValue = JSON.stringify(Object.getPrototypeOf(obj));
+			if (
+				protoValue &&
+				!(
+					protoValue === "null" ||
+					protoValue === "{}" ||
+					protoValue === "[]" ||
+					protoValue === '""' ||
+					protoValue === "false" ||
+					protoValue === "true" ||
+					protoValue === "0" ||
+					// User-defined pollution strings are whitelisted here.
+					config?.getExcludedExactMatches()?.some((str) => protoValue === str)
+				)
+			) {
+				let message;
+				if (identifier) {
+					message = `Prototype Pollution: ${identifier}.__proto__ value is ${protoValue}`;
+				} else {
+					message = `Prototype Pollution: __proto__ value is ${protoValue}`;
+				}
+				if (report) {
+					reportFinding(message);
+				} else {
+					return true;
+				}
+			}
+		} catch (e) {
+			protoValue = "";
+		}
+		// Get the same data from the object's prototype.
+		obj = Object.getPrototypeOf(obj);
```

Comment:
Is it clear that an object can't (transitively) have itself as its prototype? It sounds like it should be, but a comment acknowledging that would be helpful.

Reply:
Oh, actually it can! Good catch! We should always report a finding in that case, I think!

Reply:
After checking this, JavaScript does not allow cycles in the prototypes and will throw an Error, so, I think we're good!
 ```
Uncaught Exception: Jazzer.js: Cyclic __proto__ value
    TypeError: Cyclic __proto__ value
        at A.set __proto__ [as __proto__] (<anonymous>)
```